docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,281 @@
+# Creation and Distribution of Secrets
+
+## Status
+
+** proposed **
+
+## Context
+
+This ADR seeks to clarify and prioritize the secret handling approach taken by EdgeX.
+
+EdgeX microservices need a number of secrets to be created and distributed
+in order to create a functional, secure system.
+Among these secrets are:
+
+- Privileged administrator passwords (such as a database superuser password)
+- Service account passwords (e.g. for a database)
+- PKI certificates
+
+There is a lack of consistency on how secrets are created and distributed to EdgeX microservices,
+and when developers need to add new components to the system,
+it is unclear on what the preferred approach should be.
+
+This document assumes a threat model wherein the EdgeX services are sandboxed
+(such as in a snap or a container) and the host system is trusted,
+and all services running in a single snap share a trust boundary.
+
+### Current practices survey
+
+A survey on the existing EdgeX secrets reveals the following appoaches.
+
+A designation of "compliant" means that the current implementation
+is aligned with the recommended practices documented in the next section.
+A designation of "non-compliant" means that the current implementation
+uses an implemention mechanism outside of the recommended practices documented in the next section.
+A "non-compliant" implementation is a candidate for refactoring
+to bring the implementation into conformance with the recommended practices.
+
+- PKI private keys
+  * Docker: PKI generated by standalone utility every boot. Distribution via secrets volume. (Compliant.)
+  * Snaps: PKI generated by standalone utility every boot. Deployed to `$SNAP_DATA/secrets`. (Compliant.)
+
+- Secret store master password
+  * Docker: Distribution via persistent docker volume. (Non-compliant.)
+  * Snaps: Stored in `$SNAP_DATA/config/security-secrets-setup/res`. (Non-compliant.)
+
+- Secret store per-service authentication tokens
+  * Docker: Distribution via secrets volume, generated every boot. (Compliant.)
+  * Snaps: Distribution via secrets volume, generated every boot. (Compliant.)
+
+- Postgres superuser password
+  * Docker: Hard-coded into docker-compose file, checked in to source control. (Non-compliant.)
+  * Snaps: Deployed to `$SNAP_DATA/config/postgres/kongpw`. (Non-compliant.) 
```

Comment:
The only thing needed to make it compliant is to cache the password in Vault instead of the file system.  Will update to note exactly what part of the above flow is non-compliant.

Reply:
Sure, but doing so implies enabling PostgreSQL to read the password from Vault, which it turns out might not be all that difficult:

https://www.vaultproject.io/docs/secrets/databases/postgresql

Reply:
I was thinking something simpler.  Since the only user of Postgres is Kong, I was thinking to use "apg" to generate the password, store it in Vault, and then use it to configure Postgres at install time.  For Kong, use Andre's security-secrets-read to read the password when starting Kong, and inject it into an environment variable.

The Vault postgres database secrets engine needs a superuser password configured before it can create user accounts.  Would complicate things since it would then be necessary to generate a user account for Kong, and then install the schema under that account, and then use that account password.  I haven't experimented with it, but it is not clear to me if the "create credential" call to the database secrets engine caches the generated credential automatically, or whether it needs to be explicitly stored.  If the later, it would have just been simpler, or a single-user-DB, to not have bothered.